File: 014/e014.swift
// Longest Collatz Sequence
var memo : [Int : Int] = [1 : 0]

func f(_ n : Int) -> Int {
    return n % 2 == 0 ? n / 2 : 3 * n + 1
}

func cycleLength(_ n : Int) -> Int {
    if memo[n] == nil {
        memo[n] = cycleLength(f(n)) + 1
    }
    return memo[n]!
}

let n = 999999
let chains = [Int](1...n).map { (len: cycleLength($0), index: $0) }
let longest = chains.max { a, b in a.len < b.len }
print(longest!.index)

File: 014/e014.go
package main

import "fmt"

var memo map[int]int

func f(n int) int {
    if n%2 == 0 {
        return n / 2
    } else {
        return 3 * n + 1
    }
}

func cycleLength(n int) int {
    if _, ok := memo[n]; !ok {
        memo[n] = cycleLength(f(n)) + 1
    }
    return memo[n]
}

func main() {
    memo = make(map[int]int)
    memo[1] = 0

    N := 1000000
    chains := make([]int, N)
    for i:=1; i < N; i++ {
        chains[i] = cycleLength(i)
    }

    longest := 1
    index := 1
    for i:=1; i < N; i++ {
        if chains[i] > longest {
            longest = chains[i]
            index = i
        }
    }

    fmt.Println(index)
}
